### ISSUE TITLE
gRPC update causing failure of API calls with large responses

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#51099]
For e.g the density test started failing on all our kubemark jobs that are big (i.e on >= 100 nodes):
- https://k8s-testgrid.appspot.com/sig-scalability#kubemark-100
- https://k8s-testgrid.appspot.com/sig-scalability#kubemark-500

And it clearly seems to be a regression caused by this PR (from the diff) - https://github.com/kubernetes/kubernetes/pull/50531 - which updated gRPC vendor code. The error we're seeing is the following:

```
Expected error:
    <*errors.StatusError | 0xc42d8c0900>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {SelfLink: "", ResourceVersion: ""},
            Status: "Failure",
            Message: "rpc error: code = ResourceExhausted desc = grpc: received message larger than max (26710706 vs. 4194304)",
            Reason: "",
            Details: nil,
            Code: 500,
        },
    }
    rpc error: code = ResourceExhausted desc = grpc: received message larger than max (26710706 vs. 4194304)
not to have occurred
```

It's coming from this line of the test - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scalability/density.go#L232 - where we list all pods across all namespaces. Seems like the MTU for gRPC calls changed as a result of the update.

cc @kubernetes/sig-scalability-bugs @RenaudWasTaken @wojtek-t

**Comment by shyamjvs** [comment]
Even our 100-node gce cluster tests are failing:
- https://k8s-testgrid.appspot.com/google-gce-scale#gci-gce
- https://k8s-testgrid.appspot.com/google-gce-scale#gce

@wojtek-t This seems like a potentially dangerous change for large clusters. Is it ok to revert it?

**Comment by shyamjvs** [comment]
It's coming from here - https://github.com/kubernetes/kubernetes/blob/master/vendor/google.golang.org/grpc/server.go#L51 - where the gRPC is setting default receive size to 4MB.

**Comment by wojtek-t** [comment]
I think it's fine to revert short term.

But we should try to fix this, as we can't stick with old version of gRPC for a long time. If it's not possible to overwrite this parameter, we should work with gRPC team to make it possible.

@smarterclayton

**Comment by wojtek-t** [comment]
@xiang90 @hongchaodeng - did you face a similar issue with etcd?

**Comment by shyamjvs** [comment]
There seems to be a way to override the value for the gRPC server using - https://github.com/kubernetes/kubernetes/blob/master/vendor/google.golang.org/grpc/server.go#L187

**Comment by shyamjvs** [comment]
But since it's inside etcd vendor code, can't touch it directly.

> @xiang90 @hongchaodeng - did you face a similar issue with etcd?

I'm curious too how etcd faced this?

**Comment by xiang90** [comment]
@shyamjvs 

We need to unlimit this option for now. etcd upstream is working on supporting frag responses due to this limit.

**Comment by shyamjvs** [comment]
@xiang90 Makes sense. Any idea how to do this without touching the etcd vendor code? And why were we not having this problem before (because seems like this limit was also there before this PR)?

**Comment by vishh** [comment]
@shyamjvs @xiang90 @wojtek-t The grpc update is required for v1.8 feature. Are there any alternative solutions in addition to waiting for an ectd update?

**Comment by xiang90** [comment]
@shyamjvs The gRPC issue is introduced in gRPC 1.4. Can you try gRPC 1.3, which also supports keepalive? The current etcd vendoring is etcd3.1, which does not really work with etcd 1.4 for the reason you see here.

**Comment by shyamjvs** [comment]
Discussed offline with @RenaudWasTaken. As @xiang90 pointed out, we should try updating to 1.3 which is the min version supporting keepalive. Anything above that apparently seems to set a hard limit on the MTU which has to be configured by etcd (for that etcd would also have to be updated to 3.3).

@RenaudWasTaken is going to try updating to 1.3. Thanks everyone for the leads.

**Comment by vishh** [comment]
@RenaudWasTaken can you confirm that you are working on a fix?

**Comment by RenaudWasTaken** [comment]
@vishh yes, just started running `godep-save.sh` had some troubles setting my environment on my laptop but you should see a PR for this soon.

**Comment by xiang90** [comment]
etcd 3.3 release supports gRPC 1.4. It will be release soonish. When we bump gRPC, we need to bump etcd. 

/cc @shyamjvs @wojtek-t @jpbetz

**Comment by shyamjvs** [comment]
Reopening this issue as we're seeing this again since last friday. All our large cluster tests (>= 100 nodes) are failing with:

```
    <*errors.StatusError | 0xc423b25e60>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
            Status: "Failure",
            Message: "rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5292070 vs. 4194304)",
            Reason: "",
            Details: nil,
            Code: 500,
        },
    }
    rpc error: code = ResourceExhausted desc = grpc: received message larger than max (5292070 vs. 4194304)
```

PR https://github.com/kubernetes/kubernetes/pull/53128 updated the grpc dependency and that seems to break it.
We should revert it and bump it only when etcd is also bumped as pointed out in the above comment.

cc @dixudx @caesarxuchao @wojtek-t

**Comment by shyamjvs** [comment]
@sttts Any idea why merge commits to your personal forks show up here? Adds some noise to the thread :)

**Comment by sttts** [comment]
@shyamjvs it's were the staging repos are prepared for publishing. If PRs have links to issues/PRs, github will do this. Not much we can do.

**Comment by porridge** [comment]
/reopen
gRPC was bumped again, and so this fails again: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/31

**Comment by RenaudWasTaken** [comment]
@porridge do you think it's worth trying to prevent people from updating gRPC ?
Maybe commenting the Godeps file or adding a check in the verify-godeps.sh ?

**Comment by porridge** [comment]
If it does not cause problem with the vendoring tools, then adding a
comment pointing at this issue might help avoid lost effort I guess.

**Comment by porridge** [comment]
This is causing failures even for 100-node clusters. Example:
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/9122

**Comment by porridge** [comment]
Umbrella issue: https://github.com/kubernetes/kubernetes/issues/56062

**Comment by dims** [comment]
@shyamjvs @enisoc @jberkus @porridge we should probably move the milestone from 1.9 to 1.10

**Comment by porridge** [comment]
@dims I agree, but lack privileges

**Comment by wojtek-t** [comment]
done

**Comment by jpbetz** [comment]
This is fixed by https://github.com/coreos/etcd/pull/9047, which has just been made available in etcd v3.2.12. We've created https://github.com/kubernetes/kubernetes/pull/57480 to upgrade kubernetes to this etcd version. We'll need to scalability test this PR before merging it.
